### PR TITLE
Handle a device already being connected with the BlueZ Dbus backend

### LINKED ISSUE #992: Device may already be connected when calling connect with BlueZ
If the device is connected, and then the application/docker container is restarted and the device is not disconnected, the next connection attempt will fail because the device is already connected on the host but we think it is not connected and a connect call will fail.

```
      "/org/bluez/hci1/dev_BD_24_6F_85_AA_61": {
        "org.freedesktop.DBus.Introspectable": {},
        "org.bluez.Device1": {
          "Address": "BD:24:6F:85:AA:61",
          "AddressType": "public",
          "Name": "Dream~BD246F85AA61",
          "Alias": "Dream~BD246F85AA61",
          "Appearance": 962,
          "Icon": "input-mouse",
          "Paired": false,
          "Trusted": false,
          "Blocked": false,
          "LegacyPairing": false,
          "RSSI": -36,
          "Connected": true,
          "UUIDs": [
            "00001800-0000-1000-8000-00805f9b34fb",
            "00001801-0000-1000-8000-00805f9b34fb",
            "0000180a-0000-1000-8000-00805f9b34fb",
            "0000ffd0-0000-1000-8000-00805f9b34fb",
            "0000ffd5-0000-1000-8000-00805f9b34fb"
          ],
          "Modalias": "usb:v045Ep0040d0300",
          "Adapter": "/org/bluez/hci1",
          "ManufacturerData": {
            "20808": {
              "__type": "<class 'bytearray'>",
              "repr": "bytearray(b'364656')"
            }
          },
          "ServicesResolved": true
        },
        "org.freedesktop.DBus.Properties": {}
      },
```

### PR DESCRIPTION
**Pull request by bdraco**
Fixes #992

### CODE REVIEW

File: bleak/backends/bluezdbus/client.py [hunk 1]
```diff
@@ -155,16 +155,17 @@ def on_value_changed(char_path: str, value: bytes) -> None:
...
         try:
             try:
-                async with async_timeout.timeout(timeout):
-                    reply = await self._bus.call(
-                        Message(
-                            destination=defs.BLUEZ_SERVICE,
-                            interface=defs.DEVICE_INTERFACE,
-                            path=self._device_path,
-                            member="Connect",
+                if not manager.is_connected(self._device_path):
```
Comment (dlech) [inline bleak/backends/bluezdbus/client.py:158]: Let's add a comment explaining why we need this in the BlueZ backend (i.e. to make it behave more like other backends and to work around https://github.com/bluez/bluez/issues/89).

**Review by dlech — COMMENTED** [review]
Please add a changelog entry too.

**Comment by dlech** [comment]
For the record, there are a couple of pitfalls with this:

1. This will now allow disconnecting devices that we don't "own" the connection to. If the device is used by the OS (e.g. a keyboard or mouse), this could be problematic.
2. Many devices that are already connected won't be advertising, so may not be discoverable with BleakScanner so additional changes like those suggested in https://github.com/hbldh/bleak/issues/992#issuecomment-1242760927 may be needed to fully make use of this change.

In spite of these, this seems like a step in the right direction.

**Comment by bdraco** [comment]
> 2\. Many devices that are already connected won't be advertising, so may not be discoverable with BleakScanner so additional changes like those suggested in [Device may already be connected when calling connect with BlueZ #992 (comment)](https://github.com/hbldh/bleak/issues/992#issuecomment-1242760927) may be needed to fully make use of this change.

The workaround I'm using for that is https://github.com/Bluetooth-Devices/bleak-retry-connector/blob/e8bb72257a4e58a55a839525b2762d7db8dcab0c/src/bleak_retry_connector/__init__.py#L246

Note: its not so pretty and I'd like to get rid of it

**Comment by bdraco** [comment]
This also need to handle the device not being in the history anymore
```
2022-09-10 13:07:42.089 ERROR (MainThread) [aiohomekit.utils] Failure running background task: Task-10942
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/aiohomekit/utils.py", line 29, in _handle_task_result
    task.result()
  File "/usr/local/lib/python3.10/site-packages/aiohomekit/controller/ble/pairing.py", line 411, in _process_disconnected_events
    results = await self.get_characteristics(list(self.subscriptions))
  File "/usr/local/lib/python3.10/site-packages/aiohomekit/controller/ble/client.py", line 75, in _async_wrap
    return await func(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/aiohomekit/controller/ble/pairing.py", line 707, in get_characteristics
    return await self._get_characteristics_without_retry(characteristics)
  File "/usr/local/lib/python3.10/site-packages/aiohomekit/controller/ble/pairing.py", line 111, in _async_wrap
    return await func(self, *args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/aiohomekit/controller/ble/pairing.py", line 714, in _get_characteristics_without_retry
    await self._populate_accessories_and_characteristics()
  File "/usr/local/lib/python3.10/site-packages/aiohomekit/controller/ble/pairing.py", line 583, in _populate_accessories_and_characteristics
    await self._ensure_connected()
  File "/usr/local/lib/python3.10/site-packages/aiohomekit/controller/ble/pairing.py", line 323, in _ensure_connected
    self.client = await establish_connection(
  File "/usr/local/lib/python3.10/site-packages/aiohomekit/controller/ble/connection.py", line 48, in establish_connection
    return await retry_establish_connection(
  File "/usr/local/lib/python3.10/site-packages/bleak_retry_connector/__init__.py", line 410, in establish_connection
    await client.connect(
  File "/usr/local/lib/python3.10/site-packages/bleak_retry_connector/__init__.py", line 138, in connect
    connected = await super().connect(
  File "/usr/local/lib/python3.10/site-packages/bleak/backends/bluezdbus/client.py", line 158, in connect
    if not manager.is_connected(self._device_path):
  File "/usr/local/lib/python3.10/site-packages/bleak/backends/bluezdbus/manager.py", line 610, in is_connected
    return self._properties[device_path][defs.DEVICE_INTERFACE]["Connected"]
KeyError: 'org.bluez.Device1'
```

**Comment by bdraco** [comment]
Still doing testing. Will do another turn once done testing

File: bleak/backends/bluezdbus/client.py (hunk 1 above)
Comment (bdraco) [reply bleak/backends/bluezdbus/client.py:166]: a4ed7f8cfdea2752ed4ff14957ebf503be293d66

**Comment by bdraco** [comment]
Additional testing looks good

File: bleak/backends/bluezdbus/manager.py [hunk 2]
```diff
@@ -597,6 +597,22 @@ def get_device_name(self, device_path: str) -> str:
...
+            device_path: The D-Bus object path of the device.
+
+        Returns:
+            The current property value.
+        """
+        return bool(
+            device_path in self._properties
+            and defs.DEVICE_INTERFACE in self._properties[device_path]
+            and self._properties[device_path][defs.DEVICE_INTERFACE]["Connected"]
+        )
```
Comment (dlech) [inline bleak/backends/bluezdbus/manager.py:614]: perhaps a bit simpler/easier to read?

```suggestion
        try:
            return self._properties[device_path][defs.DEVICE_INTERFACE]["Connected"]
        except KeyError:
            return False
```

Comment (bdraco) [reply bleak/backends/bluezdbus/manager.py:614]: I'm probably over thinking it but we don't get zero cost exception handling until python 3.11
https://github.com/python/cpython/issues/84403

File: CHANGELOG.rst [hunk 3]
```diff
@@ -20,6 +20,7 @@ Changed
   To use the cache, call ``connect`` and ``get_services`` with the ``dangerous_use_bleak_cache``
   argument to avoid services being resolved again.
 * The BlueZ D-Bus backend now uses ``dbus-fast`` instead of ``dbus-next`` which significantly improves performance.
+* The BlueZ D-Bus backend will not avoid trying to connect to devices that are already connected.
```
Comment (dlech) [inline CHANGELOG.rst:23]:
```suggestion
* The BlueZ D-Bus backend will not avoid trying to connect to devices that are already connected. Fixes #992.
```

**Pull request merged**